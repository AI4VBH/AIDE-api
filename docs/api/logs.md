<!--
  ~ Copyright 2022 Guy’s and St Thomas’ NHS Foundation Trust
  ~
  ~ Licensed under the Apache License, Version 2.0 (the "License");
  ~ you may not use this file except in compliance with the License.
  ~ You may obtain a copy of the License at
  ~
  ~ http://www.apache.org/licenses/LICENSE-2.0
  ~
  ~ Unless required by applicable law or agreed to in writing, software
  ~ distributed under the License is distributed on an "AS IS" BASIS,
  ~ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  ~ See the License for the specific language governing permissions and
  ~ limitations under the License.
-->

# Logs

Returns logs from Elastic.

## `GET /logs/:id`

Returns logs for a given execution ID.

### Parameters

| Parameter Type | Label | Description |
|----------------|-------|-------------|
| URL | `:id` | `UUID` Task execution id. |

### Responses

Returns an array of `LogsDto`.

| Code | Description |
|------|-------------|
| `200` | Logs for a given task execution id retrieved. |
| `400` | Missing task execution id |
| `500` | Server error. The response will be a [Problem details](https://datatracker.ietf.org/doc/html/rfc7807) object with server error details. |

### Example Request

```bash
curl --location --request GET 'http://localhost:5000/logs/18119a95-b097-498d-a815-62c140e201e7'
```

### Example Responses

```json
[
  {
    "level": "Information",
    "renderedMessage": "Task dispatch event saved \"04a0cded-6aad-408c-b45d-1852ad54fae3\".",
    "timestamp": "2022-09-26T15:19:23.6849023Z",
  },
  {
    "level": "Information",
    "renderedMessage": "Publishing message to \"rabbitmq.monai\"/\"monaideploy\". Exchange=\"monaideploy\", Routing Key=\"md.tasks.dispatch\".",
    "timestamp": "2022-09-26T15:19:22.9668840Z",
  }
]
```