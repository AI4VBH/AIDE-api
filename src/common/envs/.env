PORT=5000

KEYCLOAK_URL='http://0.0.0.0:8443'
KEYCLOAK_MAIN_REALM='master'
KEYCLOAK_REALM='Demo-Realm'
KEYCLOAK_CLIENTID='nest-app'
KEYCLOAK_SECRET='83790b4f-48cd-4b6c-ac60-451a918be4b9'
KEYCLOAK_STATIC_ROLES='["admin", "clinician"]'

DATABASE_HOST='localhost'
DATABASE_NAME='nest_api'
DATABASE_USER='postgres'
DATABASE_PASSWORD='admin'
DATABASE_PORT=5432

ELASTIC_HOST=''
ELASTIC_PORT=''
ELASTIC_USE_SSL=''
ELASTIC_USERNAME=''
ELASTIC_PASSWORD=''
ELASTIC_INDEX=''

MONAI_API_HOST='https://localhost:8081/'
MIG_API_HOST='http://localhost:5010/'

MINIO_HOST='localhost'
MINIO_PORT=9001
MINIO_USE_SSL=false
MINIO_BUCKET='monaideploy'
MINIO_ACCESS_KEY='wOzmbchico52AMp3'
MINIO_SECRET_KEY='zyoT9SiMLiTCvNT0DS636FMiN4yi2258'
