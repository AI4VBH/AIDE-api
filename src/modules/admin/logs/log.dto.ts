export class LogsDTO {
  json: LogDTO;

  public static from(dto: Partial<LogsDTO>) {
    const it = new LogsDTO();
    it.json = dto.json;
    return it;
  }

  // TODO: Implement fromEntity method
  // public static fromEntity(entity: Logs) {
  //   return this.from({
  //     task_id: entity.id,
  //   });
  // }

  // TODO: Implement toEntity method
  // public toEntity(logs: Logs = null) {
  //   const it = new Logs();
  //   it.task_id = this.task_id;
  //   return it;
  // }
}

export class LogDTO {
  execution_id: string;
  level: string;
  line_no: number;
  logger: string;
  model_name: string;
  model_version: string;
  module: string;
  msg: string;
  thread: string;
  type: string;
  written_at: string;
  written_ts: number;

  public static from(dto: Partial<LogDTO>) {
    const it = new LogDTO();
    it.execution_id = dto.execution_id;
    it.level = dto.level;
    it.line_no = dto.line_no;
    it.logger = dto.logger;
    it.model_name = dto.model_name;
    it.model_version = dto.model_version;
    it.module = dto.module;
    it.msg = dto.msg;
    it.thread = dto.thread;
    it.type = dto.type;
    it.written_at = dto.written_at;
    it.written_ts = dto.written_ts;
    return it;
  }

  // TODO: Implement fromEntity method
  // public static fromEntity(entity: Log) {
  //   return this.from({
  //     task_id: entity.id,
  //   });
  // }

  // TODO: Implement toEntity method
  // public toEntity(log: Log = null) {
  //   const it = new Log();
  //   it.task_id = this.task_id;
  //   return it;
  // }
}
