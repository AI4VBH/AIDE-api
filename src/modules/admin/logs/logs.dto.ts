export class LogsDto {
  public renderedMessage: string;
  public timestamp: Date;
  public level: string;
}
