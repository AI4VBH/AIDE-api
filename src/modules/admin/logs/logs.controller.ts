/*
 * Copyright 2022 Guy’s and St Thomas’ NHS Foundation Trust
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

import {
  BadRequestException,
  Controller,
  Get,
  Param,
  UseFilters,
} from '@nestjs/common';
import ExternalServerExceptionFilter from 'shared/http/external-server-exception.filter';
import { LogsService } from './logs.service';

@Controller('logs')
@UseFilters(ExternalServerExceptionFilter)
export class LogsController {
  constructor(private readonly appService: LogsService) {}

  @Get(':id')
  getLogs(@Param('id') id: string) {
    if (!id) {
      throw new BadRequestException('id param cannot be empty');
    }
    return this.appService.getLogByTask(id);
  }
}
