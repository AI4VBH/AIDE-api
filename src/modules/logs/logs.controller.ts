import { Controller, Get } from '@nestjs/common';
import { Log } from './log.interface';
import { LogsService } from './logs.service';

@Controller('logs')
export class LogsController {
  constructor(private readonly appService: LogsService) {}

  @Get()
  getHello(): Log {
    return this.appService.getLog();
  }
}
