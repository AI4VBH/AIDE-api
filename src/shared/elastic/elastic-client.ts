import { Inject, Injectable, UseFilters } from '@nestjs/common';
import { ConfigService } from '@nestjs/config';
import { Client, ClientOptions } from '@opensearch-project/opensearch';
import { parseBoolean } from 'shared/util/parseBoolean';

@Injectable()
export class ElasticClient extends Client {
  searchIndex: string;
  public static createNode(config: ConfigService): string {
    const ssl = parseBoolean(config.get('ELASTIC_USE_SSL'));
    const host = config.get<string>('ELASTIC_HOST');
    const port = config.get('ELASTIC_PORT');
    const prefix = ssl ? 'https' : 'http';

    return `${prefix}://${host}:${port}`;
  }

  constructor(@Inject(ConfigService) config: ConfigService) {
    super({
      node: ElasticClient.createNode(config),
      auth: {
        username: config.get<string>('ELASTIC_USERNAME'),
        password: config.get<string>('ELASTIC_PASSWORD'),
      },
    } as ClientOptions);
    this.searchIndex = config.get<string>('ELASTIC_INDEX');
  }

  public searchWorkflowInstance(id: string) {
    return { query: { match: { 'Properties.workflowInstanceId': id } } };
  }

  public searchCorrilation(id: string) {
    return { query: { match: { 'Properties.correlationId': id } } };
  }

  public searchExecution(id: string) {
    return { query: { match: { 'Properties.executionId': id } } };
  }

  public searchTask(id: string) {
    return { query: { match: { 'Properties.taskId': id } } };
  }

  public async getLogs(id: string) {
    try {
      return await this.search({
        index: this.searchIndex,
        body: this.searchExecution(id),
      });
    } catch (error) {
      return { ...error.meta };
    }
  }
}
