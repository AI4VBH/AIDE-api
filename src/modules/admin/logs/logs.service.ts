import { Inject, Injectable } from '@nestjs/common';
import { ElasticClient } from 'shared/elastic/elastic-client';
import { LogsDto } from './logs.dto';
import { IElasticLogObject } from './models/logs.interfaces';

@Injectable()
export class LogsService {
  @Inject(ElasticClient)
  private readonly elasticClient: ElasticClient;

  async getLogByTask(id: string): Promise<LogsDto[]> {
    const response = await this.elasticClient.getLogs(id);

    if (response.statusCode != 200 || response.body.hits.total.value == 0) {
      return [];
    }

    const body = response.body as IElasticLogObject;

    const dtoArr: LogsDto[] = [];
    for (const hit of body.hits.hits) {
      dtoArr.push({
        level: hit._source.Level,
        renderedMessage: hit._source.RenderedMessage,
        timestamp: hit._source.Timestamp,
      });
    }

    return dtoArr;
  }
}
