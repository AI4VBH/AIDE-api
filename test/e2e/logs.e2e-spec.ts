import { Test, TestingModule } from '@nestjs/testing';
import { INestApplication } from '@nestjs/common';
import { HttpModule } from '@nestjs/axios';
import { ConfigModule } from '@nestjs/config';
import * as request from 'supertest';
import { LogsController } from 'modules/admin/logs/logs.controller';
import { LogsService } from 'modules/admin/logs/logs.service';
import { rest } from 'msw';
import { setupServer } from 'msw/node';
import { HttpConfigService } from 'shared/http/http.service';
import logMocks from '../test_data/mocks/logs/logs-index';
import { ElasticClient } from 'shared/elastic/elastic-client';

const server = setupServer();
const mockElasticHost = 'localhost';
const mockElasticPort = '7339';
const mockElasticUseSsl = 'true';
const mockElasticUsername = 'username';
const mockElasticPassword = 'password';
const mockElasticIndex = 'logstash*';
const mockHttp = `https://${mockElasticHost}:${mockElasticPort}`;

describe('/Logs Integration Tests', () => {
  let app: INestApplication;

  beforeAll(() => {
    server.listen({
      onUnhandledRequest: 'bypass',
    });
    server.printHandlers();
  });

  afterAll(() => {
    server.close();
  });

  beforeEach(async () => {
    server.resetHandlers();

    const moduleFixture: TestingModule = await Test.createTestingModule({
      imports: [
        ConfigModule.forRoot({
          isGlobal: true,
          load: [
            () => ({
              ELASTIC_HOST: mockElasticHost,
              ELASTIC_PORT: mockElasticPort,
              ELASTIC_USE_SSL: mockElasticUseSsl,
              ELASTIC_USERNAME: mockElasticUsername,
              ELASTIC_PASSWORD: mockElasticPassword,
              ELASTIC_INDEX: mockElasticIndex,
            }),
          ],
        }),
        HttpModule.registerAsync({
          useClass: HttpConfigService,
        }),
      ],
      controllers: [LogsController],
      providers: [LogsService, ElasticClient],
    }).compile();

    app = moduleFixture.createNestApplication();
    await app.init();
  });

  it.each([logMocks.basicLogs1, logMocks.basicLogs2, logMocks.basicLogs3])(
    '(GET) /logs/:taskid with returned data',
    async (payload) => {
      server.use(
        rest.post(
          `${mockHttp}/${mockElasticIndex}*/_search`,
          (_request, response, context) => {
            return response(context.json(payload));
          },
        ),
      );
      const response = await request(app.getHttpServer()).get(
        `/logs/04a0cded-6aad-408c-b45d-1852ad54fae3`,
      );
      expect(response.body).toMatchSnapshot();
      expect(response.statusCode).toBe(200);
    },
  );

  it('(GET) /logs/:taskid without returned data', async () => {
    server.use(
      rest.post(
        `${mockHttp}/${mockElasticIndex}*/_search`,
        (request, response, context) => {
          return response(context.json(logMocks.emptyLogData));
        },
      ),
    );
    const response = await request(app.getHttpServer()).get(
      '/logs/ef8dd042-24f6-4dc5-9797-5a55e2492063',
    );
    expect(response.body).toMatchObject([]);
    expect(response.statusCode).toBe(200);
  });

  it('(GET) /logs/:taskid with unparsable query', async () => {
    server.use(
      rest.post(
        `${mockHttp}/${mockElasticIndex}*/_search`,
        (request, response, context) => {
          return response(
            context.status(400),
            context.json(logMocks.unparsableQueryError),
          );
        },
      ),
    );
    const response = await request(app.getHttpServer()).get(
      '/logs/6d4a5567-d87a-4784-8078-954ec2d9bd68',
    );
    expect(response.body).toMatchObject([]);
    expect(response.statusCode).toBe(200);
  });

  it.each([408, 500, 501, 502, 503, 504])(
    '(GET) /logs/:taskid when Elastic gives general error',
    async (code) => {
      server.use(
        rest.post(
          `${mockHttp}/${mockElasticIndex}*/_search`,
          (request, response, context) => {
            return response(context.status(code));
          },
        ),
      );
      const response = await request(app.getHttpServer()).get(
        '/logs/0b0c725b-5700-49e2-b733-81759801c935',
      );
      expect(response.body).toMatchObject([]);
      expect(response.statusCode).toBe(200);
    },
  );
});
