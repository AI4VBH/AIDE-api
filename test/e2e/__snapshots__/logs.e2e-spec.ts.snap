// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/Logs Integration Tests (GET) /logs/:taskid with returned data 1`] = `
[
  {
    "level": "Information",
    "renderedMessage": "Task dispatch event saved "04a0cded-6aad-408c-b45d-1852ad54fae3".",
    "timestamp": "2022-09-26T15:19:23.6849023Z",
  },
  {
    "level": "Information",
    "renderedMessage": "Publishing message to "rabbitmq.monai"/"monaideploy". Exchange="monaideploy", Routing Key="md.tasks.dispatch".",
    "timestamp": "2022-09-26T15:19:22.9668840Z",
  },
  {
    "level": "Debug",
    "renderedMessage": "Generating Argo workflow template.",
    "timestamp": "2022-09-26T15:19:28.2563475Z",
  },
  {
    "level": "Debug",
    "renderedMessage": "Creating Argo client with base URL: "https://argo-server.argo:2746".",
    "timestamp": "2022-09-26T15:19:28.2619000Z",
  },
  {
    "level": "Information",
    "renderedMessage": "Start processing HTTP request "GET" https://argo-server.argo:2746/api/v1/workflow-templates/argo/simple-workflow",
    "timestamp": "2022-09-26T15:19:28.2769858Z",
  },
  {
    "level": "Information",
    "renderedMessage": "Sending HTTP request "GET" https://argo-server.argo:2746/api/v1/workflow-templates/argo/simple-workflow",
    "timestamp": "2022-09-26T15:19:28.2790058Z",
  },
  {
    "level": "Information",
    "renderedMessage": "Received HTTP response headers after 386.078ms - 200",
    "timestamp": "2022-09-26T15:19:28.6713753Z",
  },
  {
    "level": "Information",
    "renderedMessage": "End processing HTTP request after 398.1257ms - 200",
    "timestamp": "2022-09-26T15:19:28.6737060Z",
  },
  {
    "level": "Debug",
    "renderedMessage": "Creating Kubernetes client: host="https://10.233.0.1:443/", namespace="monai".",
    "timestamp": "2022-09-26T15:19:28.8045840Z",
  },
  {
    "level": "Debug",
    "renderedMessage": "Generating Kubernetes secrets for accessing artifacts: "argo-task-1".",
    "timestamp": "2022-09-26T15:19:28.8112027Z",
  },
]
`;

exports[`/Logs Integration Tests (GET) /logs/:taskid with returned data 2`] = `
[
  {
    "level": "Information",
    "renderedMessage": "Task dispatch event saved "04a0cded-6aad-408c-b45d-1852ad54fae3".",
    "timestamp": "2022-09-26T15:19:23.6849023Z",
  },
]
`;

exports[`/Logs Integration Tests (GET) /logs/:taskid with returned data 3`] = `
[
  {
    "level": "Trace",
    "renderedMessage": "Task dispatch event saved "04a0cded-6aad-408c-b45d-1852ad54fae3".",
    "timestamp": "2022-09-26T15:19:23.6849023Z",
  },
  {
    "level": "Debug",
    "renderedMessage": "Task dispatch event saved "04a0cded-6aad-408c-b45d-1852ad54fae3".",
    "timestamp": "2000-09-26T15:19:23.6849023Z",
  },
  {
    "level": "Information",
    "renderedMessage": "Task dispatch event saved "04a0cded-6aad-408c-b45d-1852ad54fae3".",
    "timestamp": "2022-09-26T15:19:23.6849023Z",
  },
  {
    "level": "Warn",
    "renderedMessage": "Task dispatch event saved "04a0cded-6aad-408c-b45d-1852ad54fae3".",
    "timestamp": "2022-09-26T15:37:23.6849023Z",
  },
  {
    "level": "Error",
    "renderedMessage": "Task dispatch event saved "04a0cded-6aad-408c-b45d-1852ad54fae3".",
    "timestamp": "2100-09-26T15:19:23.6849023Z",
  },
  {
    "level": "Fatal",
    "renderedMessage": "Task dispatch event saved "04a0cded-6aad-408c-b45d-1852ad54fae3".",
    "timestamp": "2022-09-26T15:19:23.6849023Z",
  },
]
`;
